src/apps/answers/service.py

```diff
@@ -1935,6 +1938,55 @@ async def get_activity_and_flow_ids_by_source_subject(self, source_subject_id: u
         """
         return await AnswersCRUD(self.answer_session).get_activity_and_flow_ids_by_source_subject(source_subject_id)
 
+    async def _filter_out_soft_deleted_subjects(
+        self, submissions_target: list[dict], submissions_respondent: list[dict]
+    ) -> set[uuid.UUID]:
+        """
+        Filter out soft-deleted source subjects from submissions_target
+        and soft-deleted target subjects from submissions_respondent
+        """
+        all_submissions = submissions_target + submissions_respondent
+
+        subject_ids = set([activityOrFlow["subject_id"] for activityOrFlow in all_submissions])
```

Comment:
#### Part 2 – `src/apps/answers/service.py`

First, update `_filter_out_soft_deleted_subjects` to be more generic:

```py
    async def _filter_out_soft_deleted_subjects(self, submissions: list[dict]) -> set[uuid.UUID]:
        """
        Filter out submissions whose subject_id column corresponds to soft-deleted subjects
        """
        subject_ids = set([activityOrFlow["subject_id"] for activityOrFlow in submissions])

        assert self.user_id
        existing_subjects = await SubjectsService(self.session, self.user_id).get_by_ids(list(subject_ids))
        existing_subject_ids = {subject.id for subject in existing_subjects}

        return existing_subject_ids
```

This change will require updating [line 1966](https://github.com/ChildMindInstitute/mindlogger-backend-refactor/pull/1669/files#diff-996adb73dc636ce745207e625c7de8b9a1560d5b1e6b17cef8f3d4f5e69020edR1966) to:

```py
        existing_subject_ids = await self._filter_out_soft_deleted_subjects(submissions_target + submissions_respondent)
```

And now, we can use `_filter_out_soft_deleted_subjects` in these two functions:

```py
    async def get_activity_and_flow_ids_by_target_subject(self, target_subject_id: uuid.UUID) -> list[dict]:
        """
        Get a list of activity and flow IDs based on answers submitted for a target subject.
        Excludes answers whose source subject was soft-deleted.

        The data returned is just a combined list of activity and flow IDs, without any
        distinction between the two.
        """
        results = await AnswersCRUD(self.answer_session).get_activity_and_flow_ids_by_target_subject(target_subject_id)
        existing_subject_ids = await self._filter_out_soft_deleted_subjects(results)
        activity_ids = [result["activity_id"] for result in results if result["subject_id"] in existing_subject_ids]

        return activity_ids

    async def get_activity_and_flow_ids_by_source_subject(self, source_subject_id: uuid.UUID) -> list[dict]:
        """
        Get a list of activity and flow IDs based on answers submitted for a source subject.
        Excludes answers whose target subject was soft-deleted.

        The data returned is just a combined list of activity and flow IDs, without any
        distinction between the two.
        """
        results = await AnswersCRUD(self.answer_session).get_activity_and_flow_ids_by_source_subject(source_subject_id)
        existing_subject_ids = await self._filter_out_soft_deleted_subjects(results)
        activity_ids = [result["activity_id"] for result in results if result["subject_id"] in existing_subject_ids]

        return activity_ids
```

That's it! Thank you! 🙏🏻🙏🏻🙏🏻

Reply:
done